server/storage/wal/wal_test.go

```diff
@@ -639,7 +639,9 @@ func TestOpenWithMaxIndex(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer w.Close()
+	defer func() {
+		w.Close()
+	}()
```

Comment:
I'm not sure this changes anything, deferred function will still refer to variable `w` which is overridden in the same scope. 
Possible fixes:
* Use different scopes
* Use different names
* Copy 

```suggestion
	defer func(w *wal.WAL) {
		w.Close()
	}(w)
```

Reply:
The proposed change looks good, but unfortunately it doesn't work. The change in this PR is working, its purpose is to make sure the `w` isn't evaluated until the function is actually called. The code you proposed will make the w to be evaluated immediately. 

Note that the w is opened two times, and also closed two times in `defer`. The reference to `w` in the first `defer` isn't valid anymore after recreating/opening the WAL. 

Please consider to update this PR per my comment above https://github.com/etcd-io/etcd/pull/14421#discussion_r962522454

Reply:
> The proposed change looks good, but unfortunately it doesn't work.

I'm super confused, how both can be true :P

I think I understood the problem. I missed `w.Close()` on line 650. 
There are 2 changes I would suggest:
* add `defer w.Close()` on after line 655 (like @ahrtr suggested)
* Renaming `w` to reflect two separate opening of raft `w1`, `w2`. Sharing same name of variable that needs to be cleanup makes code unnecessarily complicated.

Reply:
> I'm super confused, how both can be true :P


Ah:), I wanted to say your proposed change looks good in generic, but unfortunately it doesn't work in this case. :) 

I like the idea of w1 and w2 proposal, because it's clearer.

Reply:
@ahrtr I like the idea to keep defer as close to the creation as possible. But if the Create method returns error, w might be nil and defer w.Close() will cause panic. It seems quite a common pattern create -> check error -> defer. So I believe that the current location of defer is perfect.

@serathius @ahrtr renaming to w1 w2 was the first idea that came to my mind, but there would be an issue with w1. We need to explicitly close it (line 650). After that defer w1.Close() doesn't make any sense and will Close a closed object. I'm pretty much sure that Close is idempotent at the moment, but who knows whether it will always be true.

I agree that this approach is more readable, so if you guys think that double close is not a problem, I will change the PR.

Reply:
Does the following change make sense to you? @vsvastey If yes, please feel free to update this PR. thx
```
$ git diff
diff --git a/server/storage/wal/wal_test.go b/server/storage/wal/wal_test.go
index 5696f8f71..7135233b3 100644
--- a/server/storage/wal/wal_test.go
+++ b/server/storage/wal/wal_test.go
@@ -639,7 +639,6 @@ func TestOpenWithMaxIndex(t *testing.T) {
        if err != nil {
                t.Fatal(err)
        }
-       defer w.Close()
 
        es := []raftpb.Entry{{Index: uint64(math.MaxInt64)}}
        if err = w.Save(raftpb.HardState{}, es); err != nil {
@@ -651,6 +650,8 @@ func TestOpenWithMaxIndex(t *testing.T) {
        if err != nil {
                t.Fatal(err)
        }
+       defer w.Close()
+
        _, _, _, err = w.ReadAll()
        if err == nil || err != ErrSliceOutOfRange {
                t.Fatalf("err = %v, want ErrSliceOutOfRange", err)
```

Reply:
@ahrtr unfortunately, if we remove the first defer and any panic happens in range 640 - 649 or for example on w.Save at 647 returns error, then we will not close the object and probably leak resources.

Reply:
You are right. How about this? 

```
$ git diff
diff --git a/server/storage/wal/wal_test.go b/server/storage/wal/wal_test.go
index 5696f8f71..83f0dd7d5 100644
--- a/server/storage/wal/wal_test.go
+++ b/server/storage/wal/wal_test.go
@@ -635,23 +635,31 @@ func TestOpenForRead(t *testing.T) {
 func TestOpenWithMaxIndex(t *testing.T) {
        p := t.TempDir()
        // create WAL
-       w, err := Create(zaptest.NewLogger(t), p, nil)
+       w1, err := Create(zaptest.NewLogger(t), p, nil)
        if err != nil {
                t.Fatal(err)
        }
-       defer w.Close()
+       defer func() {
+               if w1 != nil {
+                       w1.Close()
+               }
+       }()
 
        es := []raftpb.Entry{{Index: uint64(math.MaxInt64)}}
-       if err = w.Save(raftpb.HardState{}, es); err != nil {
+       if err = w1.Save(raftpb.HardState{}, es); err != nil {
                t.Fatal(err)
        }
-       w.Close()
 
-       w, err = Open(zaptest.NewLogger(t), p, walpb.Snapshot{})
+       w1.Close()
+       w1 = nil
+
+       w2, err := Open(zaptest.NewLogger(t), p, walpb.Snapshot{})
        if err != nil {
                t.Fatal(err)
        }
-       _, _, _, err = w.ReadAll()
+       defer w2.Close()
+
+       _, _, _, err = w2.ReadAll()
        if err == nil || err != ErrSliceOutOfRange {
                t.Fatalf("err = %v, want ErrSliceOutOfRange", err)
        }
```

Reply:
Please feel free to update this PR if it makes sense to you, thx

Reply:
@ahrtr yes, I guess this solution is the best having both efficiency and readability. I've update the PR with it. Thanks.
Please note that I've also slightly changed a condition on line 662